compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs

```diff
@@ -534,13 +537,95 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     err.span_label(sp, format!("expected `{}` because of return type", expected));
+                    self.try_suggest_return_impl_trait(err, expected, ty, fn_id);
                     return true;
                 }
                 false
             }
         }
     }
 
+    /// check whether the return type is a generic type with a trait bound
+    /// only suggest this if the generic param is not present in the arguments
+    /// if this is true, hint them towards changing the return type to `impl Trait`
+    /// ```
+    /// fn cant_name_it<T: Fn() -> u32>() -> T {
+    ///     || 3
+    /// }
+    /// ```
+    fn try_suggest_return_impl_trait(
+        &self,
+        err: &mut DiagnosticBuilder<'_>,
+        expected: Ty<'tcx>,
+        found: Ty<'tcx>,
+        fn_id: hir::HirId,
+    ) {
+        debug!("try_suggest_return_impl_trait, expected = {:?}, found = {:?}", expected, found);
+
+        let ty_param = match expected.kind() {
+            ty::Param(ty_param) => ty_param,
+            _ => return,
+        };
+
+        let fn_node = self.tcx.hir().find(fn_id);
+
+        if let Some(hir::Node::Item(hir::Item {
+            kind:
+                hir::ItemKind::Fn(
+                    hir::FnSig { decl: hir::FnDecl { inputs, output, .. }, .. },
+                    hir::Generics { params, where_clause, .. },
+                    _body_id,
+                ),
+            ..
+        })) = fn_node
+        {
+            if let Some(generic_param) = params.get(ty_param.index as usize) {
+                let all_bounds = generic_param
+                    .bounds
+                    .iter()
+                    .chain(
+                        where_clause
+                            .predicates
+                            .iter()
+                            .flat_map(|p| match p {
+                                WherePredicate::BoundPredicate(hir::WhereBoundPredicate {
+                                    bounds,
+                                    ..
+                                }) => Some(bounds),
+                                _ => None,
+                            })
+                            .map(|bounds| bounds.iter())
+                            .flatten(),
+                    )
+                    .flat_map(|bound| match bound {
+                        GenericBound::Trait(_, _) => {
+                            self.tcx.sess.source_map().span_to_snippet(bound.span()).ok()
+                        }
+                        _ => None,
+                    })
+                    .collect::<Vec<String>>();
+                let all_bounds_str = all_bounds.join(" + ");
+                if all_bounds.len() != 0 {
+                    if inputs.iter().any(|param| {
+                        let ty = <dyn AstConv<'_>>::ast_ty_to_ty(self, param);
+                        matches!(ty.kind(), ty::Param(fn_param_ty_param) if ty_param == fn_param_ty_param)
```

Comment:
What is this checking? Why is it different than the case below?

Reply:
I changed the naming and added a comment, I hope that will clear it up